[RFC 630] add readiness healthchecks for apps (#630)
cloudfoundry/community: toc/rfc/rfc-draft-readiness-healthchecks.md

application healthcheck detects that an app instance (AI) is unhealthy, then
Diego will stop the AI, delete the AI, and reschedule a new AI.

This is too aggressive from some apps. There could be many reasons why a single
request could fail, but the app is actually running fine. Additionally, many
applications have a warm up period where they are not ready to receive requests
immediately. For example, apps might need to populate caches, load data, or wait
for external services before they mark themselves as routable. In these cases,
the app should be kept alive, but in a non-routable state.

## Proposal

### Summary
We intend to support readiness healthchecks. (This was requested previously in
this [issue](https://github.com/cloudfoundry/cloud_controller_ng/issues/1706).)
This would be an additional healthcheck that app developers could configure.
When the readiness healthcheck passes, the app is marked "ready" and the app
will be routable. When the readiness healthcheck fails, the app is marked as
"not ready" and its route will be removed from gorouter's route table.
This new readiness healthcheck will give users a healthcheck option that is less

> drastic than the current option.

Reviewer: Thanks!

### Architecture Overview
This feature will require changes in the following releases

* CF CLI
* Cloud Controller
* Diego
* Routing

1. The cloud controller will store this new data, before passing it onto the BBS
   as part of the desired LRP.
2. The Diego executor will see these new readiness healthchecks on the desired
   LRP and will run the healthchecker binary in the app container with
   configuration provided.
3. When the readiness healthcheck succeeds, the container will be marked as
   "ready". When the readiness healthcheck fails, the container will be marked
   as "not ready".
4. When the route emitter gets route information, it will inspect if the AI is
   ready or not ready. It will emit registration or unregistration messages as
   appropriate for the gorouter to consume

### CC Design

> Users will be able to set the healthcheck via the app manifest.

Reviewer: I also would vote to add the option to configure the timeout with `--timeout` in `set-health-check` as described in this [issue](https://github.com/cloudfoundry/cli/issues/2366). In my opinion It is getting more relevant to update the timeout with `set-health-check` with this new type of health check.

> Users will be able to set the healthcheck via the app manifest.

Reviewer 2: As a first iteration we can configure retrieve information about readiness health checks in the manifest. But eventually, cf CLI should provide options to configure it. 
How about push command flags: `--readiness-endpoint` and `--readiness-health-check-type`. Same for CLI commands `get-readiness-health-check` and `set-readiness-health-check`

> Users will be able to set the healthcheck via the app manifest.

Reviewer 3: Yes, still the changes to the CLI should be outlined in the RFC. We might decide to implement them at a later stage but this RFC should describe the feature as a whole as much as possible.

> Users will be able to set the healthcheck via the app manifest.

Reviewer: Sounds good and it is consistent with what we have today available in CF CLI for health check configuration.

> Users will be able to set the healthcheck via the app manifest.

Reviewer: @mariash the proposed CLI options are missing in the RFC. I think it makes sense to add them or?

> Users will be able to set the healthcheck via the app manifest.

Reviewer: This is now incorporated into the RFC. I'm resolving. Thanks @mariash for the updates!

```
applications:
- name: test-app
  processes:
  - type: web
    health-check-http-endpoint: /health
    health-check-invocation-timeout: 2
    health-check-type: http
    timeout: 80

>     readiness-health-check-http-endpoint: /health      # 👈 new property

Reviewer 3: (minor) The two checks should use different endpoints:

```suggestion
    readiness-health-check-http-endpoint: /ready       # 👈 new property
```

>   - type: web
>     health-check-http-endpoint: /health
>     health-check-invocation-timeout: 2
>     health-check-type: http
>     timeout: 80
>     readiness-health-check-http-endpoint: /health      # 👈 new property
>     readiness-health-check-invocation-timeout: 2       # 👈 new property
>     readiness-health-check-type: http                  # 👈 new property

Reviewer 4: The example is a little misleading, as it checks the same endpoint with the same timeout. Maybe we can highlight that there could be different timings or different endpoints ("liveness" via health check vs. "readiness").

```suggestion
  - type: web
    health-check-http-endpoint: /health
    health-check-invocation-timeout: 2
    health-check-type: http
    timeout: 80
    readiness-health-check-http-endpoint: /ready       # 👈 new property
    readiness-health-check-invocation-timeout: 2       # 👈 new property
    readiness-health-check-type: http                  # 👈 new property
```

```

### LRP Design

The readiness healthcheck data will be apart of the desired LRP object.

> "check_definition": {

Reviewer 3: 😉 
````suggestion
```
"check_definition": {
```
````

(but since your comment only is on the one line I can't make a proper suggestion)

    "checks": [
      {
        "http_check": {
          "port": 8080,
          "path": "/health",
          "request_timeout_ms": 10000
        },
      }
    ],
    "readiness_checks": [                  # 👈 new property
      {
        "tcp_check": {
          "port": 8080,
          "connect_timeout_ms": 10000
        },
      }

>     ],

Reviewer: - The manifest section defines `readiness-health-check-type: http` but in the data section we have `tcp_check` for the readiness health check. Is this a type?
- The readiness check data doesn't define a `path` field. Is this intended?

Reviewer 3: I thought this was intended to show that a readiness healthcheck doesn't have to be a HTTP check and can be any type of healthcheck. But it would probably be more realistic to define the healthcheck on layer 4 and the readiness check on layer 7.

Reviewer: This brings me to the point that we have to clarify in this RFC which types of health checks will be possible/allowed here because I can imagine that a `process` type of health check won't make much sense here and most probably technically not possible.

Reviewer: Thanks for the updates!

> "check_definition": {
>     "checks": [
>       {
>         "http_check": {
>           "port": 8080,
>           "path": "/health",
>           "request_timeout_ms": 10000
>         },
>       }
>     ],
>     "readiness_checks": [                  # 👈 new property
>       {
>         "tcp_check": {
>           "port": 8080,
>           "connect_timeout_ms": 10000
>         },
>       }
>     ],
>     "log_source": ""
>   },

Reviewer 4: Putting this into a code fence would be easier to read in the markdown representation.
````suggestion
```
"check_definition": {
    "checks": [
      {
        "http_check": {
          "port": 8080,
          "path": "/health",
          "request_timeout_ms": 10000
        },
      }
    ],
    "readiness_checks": [                  # 👈 new property
      {
        "tcp_check": {
          "port": 8080,
          "connect_timeout_ms": 10000
        },
      }
    ],
    "log_source": ""
  },
```
````

> ### Open Questions

Reviewer 5: Additional open question:
* How shall a rolling app deployment behave? Use the health check as today or use (if configured) the readiness check to ensure that the updated app instance is routable before an old instances is stopped.

I tend to use the readiness check (if configured).

Reviewer 2: 1. For rolling deploys, I think it makes sense to account for the application instance readiness.
2. Yes, how about routable field with values => yes/no?

Reviewer: 2. I'm fine with routable with values => yes/no

Reviewer: My questions got answered here. @stephanme can we resolve this?

Reviewer 5: An extra `routable` field in the [process stats object](https://v3-apidocs.cloudfoundry.org/index.html#the-process-stats-object) should work for the CF API. Maybe we can add this to the CC section above.

My questions are answered as well.

Reviewer 2: Updating RFC with this information.

> * What logging and metrics would be helpful for app devs and operators?

Reviewer 2: @beyhan so far we are adding log line to app logs similar to how we log when container becomes healthy/unhealthy. "Container became ready", "Container became not ready". 

For app events or metrics would be interesting to hear use cases, when event would be useful vs metric.

> * What logging and metrics would be helpful for app devs and operators?

Reviewer: We have for example an alert notification service which works based on events. I know that people configure alerts for app.crash events. I can imagine that people could be interested in configuring alerts for apps which are not routable.

Regarding the metrics I don't have any concrete use case. In the linked comment App Autoscaling was mentioned. Additionally, monitoring dashboards could be another use case where routable instances could be visualised. To my knowledge we don't have any sophisticated metrics for apps natively supported in CF. We have only the [container metrics](https://docs.cloudfoundry.org/loggregator/container-metrics.html). Usually, CF users use third party tools to collect metrics (mainly around resource usage) but I can imagine that third party tools will have hard time to generate a metric about whether an app instance is routable or not.

> * What logging and metrics would be helpful for app devs and operators?

Reviewer: The RFC has now a proposal for an event. Regarding the metrics part I don't have any concrete use case, that is why we can leave it for now. I think it makes sense to have a separate RFC which looks into possible app metrics. @mariash do you have any additional comments here?

> * What logging and metrics would be helpful for app devs and operators?

Reviewer 2: @beyhan makes sense. We can add rfc for metrics later. I am also updating the log line to clarify that instance becomes un-routable per suggestion by @ameowlia.

> * What logging and metrics would be helpful for app devs and operators?

Reviewer: Thanks!

applications have a warm up period where they are not ready to receive requests
immediately. For example, apps might need to populate caches, load data, or wait
for external services before they mark themselves as routable. In these cases,
the app should be kept alive, but in a non-routable state.

## Proposal

### Summary
We intend to support readiness healthchecks. (This was requested previously in
this [issue](https://github.com/cloudfoundry/cloud_controller_ng/issues/1706).)
This would be an additional healthcheck that app developers could configure.
When the readiness healthcheck passes, the AI is marked "ready" and the AI will
be routable. When the readiness healthcheck fails, the AI is marked as "not
ready" and its route will be removed from gorouter's route table. This new
readiness healthcheck will give users a healthcheck option that is less drastic
than the current option.

## Types of readiness healthcheck

Readiness healthcheck can be either "http" or "port" type. The format of healthcheck
type is [similar to liveness
healthcheck](https://docs.cloudfoundry.org/devguide/deploy-apps/healthchecks.html).

> The "process" healthcheck type will not be supported since it doesn't make sense

Reviewer 6: Does this need to be updated?  In the PRs for this feature, `process` is supported as an option from the API (even if it effectively means there is no readiness check)

Author: Good catch. Will update.

> Readiness healthcheck can be either "http" or "port" type. The format of healthcheck
> type is [similar to liveness
> healthcheck](https://docs.cloudfoundry.org/devguide/deploy-apps/healthchecks.html).
> The "process" healthcheck type will not be supported since it doesn't make sense
> to have "process" readiness healthcheck type. Once any defined process exits AI
> is marked as crashed.

Author: ```suggestion
[Similar to liveness healthchecks](https://docs.cloudfoundry.org/devguide/deploy-apps/healthchecks.html), readiness healthcheck can be of type "http", "port", or "process".
However, when a user selects the "process" healthcheck type, nothing will be passed to the LRP, because once a process exits the AI
is marked as crashed and Diego will attempt a restart. The default readiness healthcheck type is "process", which is backwards compatible.
```

This feature will require changes in the following releases

* CF CLI
* Cloud Controller
* Diego
* Routing

1. The cloud controller will store this new data, before passing it onto the BBS
   as part of the desired LRP.
2. The Diego executor will see these new readiness healthchecks on the desired
   LRP and will run the healthchecker binary in the app container with
   configuration provided.
3. When the readiness healthcheck succeeds, the actual LRP will be marked as
   "ready". When the readiness healthcheck fails, the actual LRP will be marked
   as "not ready".
4. When the route emitter gets route information, it will inspect if the AI is
   ready or not ready. It will emit registration or unregistration messages as
   appropriate for the gorouter to consume.

### CC Design
Users will be able to set the readiness healthcheck via the app manifest.

```
applications:
- name: test-app
  processes:
  - type: web
    health-check-http-endpoint: /health
    health-check-invocation-timeout: 2
    health-check-type: http
    timeout: 80
    readiness-health-check-http-endpoint: /ready       # 👈 new property
    readiness-health-check-invocation-timeout: 2       # 👈 new property

>     readiness-health-check-type: http                  # 👈 new property

Author: ```suggestion
    readiness-health-check-type: http                  # 👈 new property
    readiness-health-check-interval: 5                 # 👈 new property
```

with values `true` or `false` will display the routable status of the process.

### LRP Design

The readiness healthcheck data will be apart of the desired LRP object.

```json
"check_definition": {
    "checks": [
      {
        "http_check": {
          "port": 8080,
          "path": "/health",
          "request_timeout_ms": 10000
        },
      }
    ],
    "readiness_checks": [                  # 👈 new property
      {
        "tcp_check": {
          "port": 8080,
          "connect_timeout_ms": 10000
        },
      }
    ],
    "log_source": ""
  },
```

### CLI Changes

The `routable` field of the process stats API object property will be used in
CF CLI `cf app` output.

```
     state     routable   since                  cpu    memory          disk         logging            details
#0   running   yes        2023-06-27T15:07:14Z   0.6%   46.8M of 192M   179M of 1G   0/s of unlimited
#1   running   no         2023-06-27T15:11:43Z   0.0%   0 of 0          0 of 0       0/s of 0/s
#2   running   yes        2023-06-27T15:11:43Z   0.0%   0 of 192M       0 of 1G      0/s of 0/s
```

New CLI options will be added to `cf push` command:

> * `--readiness-endpoint` will allows to set the endpoint for http readiness

Author: ```suggestion
* `--readiness-endpoint` will set the endpoint for http readiness
```

  checks.

> * `--readiness-health-check-type` will allows to set the type of the readiness
>   check: "http" or "port".

Author: ```suggestion
* `--readiness-health-check-type` will set the type of the readiness
  check: "http", "port", or "process".
* `--readiness-health-check-interval` will set the interval of the readiness
  check. Must be an integer greater than 0."
```
